Fix off-by-one error checking coinbase maturity in optional UTxOs

nymius opened the pull request:
### Description

As I was developing the changes in #1798 I discover issue #1810. So I introduced the fixes in that PR but later I split them in two to ease the review by suggestion of @oleonardolima .

The `preselect_utxos` method has an off-by-one error that is making the selection of optional UTxOs too restrictive, by
requiring the coinbase outputs to surpass or equal coinbase maturity time at the current height of the selection, and
not in the block in which the transaction may be included in the blockchain.

The changes in this commit fix it by considering the maturity of the coinbase output at the spending height and not
the transaction creation height, this means, a +1 at the considered height at the moment of building the
transaction.

Fixes #1810.

### Notes to the reviewers

Tests for issue #1810 have not been explicitly added, as there already was a `text_spend_coinbase` test which was corrected to ensure coinbase maturation is considered in alignment with the new logic.

Changes are not breaking but I'm modifying slightly the documentation for the public method `TxBuilder::current_height` to adjust to the fixed code. Does this merit an entry in the CHANGELOG?

### Changelog notice

`Wallet` now considers a utxo originated from a coinbase transaction (`coinbase utxo`) as available for selection if it will mature in the next block after the height provided to the selection, the current height by default. The previous behavior allowed selecting a `coinbase utxo` only when the height at the moment of selection was equal to maturity height or greater.

### Checklists

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing
* [x] I've updated existing tests to match the fix
* [x] I've updated docs to match the fix logic
* [x] This pull request DOES NOT break the existing API
* [x] I'm linking the issue being fixed by this PR

oleonardolima reviewed (commented):
cACK efe9b4a93beb1a7a1683c2e12677b47c919ef25d

LagginTimes reviewed (approved):
ACK 03b7ecaea1176f55c6792d8798ead72b23082af7

ValuedMammal commented:
Looks good to me. I would add to Changelog something like this

`Wallet` now considers a coinbase output available for selection if it will mature in the next block after the current height; the previous behavior allowed selecting a coinbase output only when the current height reached the maturity height or greater.

nymius committed:
fix(wallet): off-by-one error checking coinbase maturity in optional UTxOs

The `preselect_utxos` method has an off-by-one error that is making the
selection of optional UTxOs too restrictive, by requiring the coinbase
outputs to surpass or equal coinbase maturity time at the current height
of the selection, and not in the block in which the transaction may be
included in the blockchain.

The changes in this commit fix it by considering the maturity of the
coinbase output at the spending height and not the transaction creation
height, this means, a +1 at the considered height at the moment of
building the transaction.

nymius commented:
> Looks good to me. I would add to Changelog something like this
> 
> `Wallet` now considers a coinbase output available for selection if it will mature in the next block after the current height; the previous behavior allowed selecting a coinbase output only when the current height reached the maturity height or greater.

I got confused and added an entry to `crates/wallet/CHANGELOG.md`. Fortunately, I checked the CHANGELOG policy again, remove it and added the comment to the CHANGELOG section of the PR header.

evanlinjin commented on crates/wallet/src/wallet/mod.rs:
```diff
@@ -2052,9 +2052,10 @@ impl Wallet {
-                                spendable &= (current_height
-                                    .saturating_sub(anchor.block_id.height))
-                                    >= COINBASE_MATURITY;
+                                let spend_height = current_height + 1;
```
Nit: I think a better name would be `inclusion_height`.

nymius replied:
Won't be needed if #1798 is merged.

evanlinjin reviewed (approved):
ACK 03b7ecaea1176f55c6792d8798ead72b23082af7
